src/sparseml/pytorch/openpifpaf/utils.py

```diff
@@ -12,8 +12,116 @@
+
+
+LOG = logging.getLogger(__name__)
 
 
 class SparseMLTrainer(openpifpaf.network.Trainer):
-    ...
+    recipe: str = None
+
+    def __init__(
+        self,
+        model: torch.nn.Module,
+        loss,
+        optimizer,
+        out,
+        num_batches_per_epoch,
+        *,
+        checkpoint_shell=None,
+        lr_scheduler=None,
+        device=None,
+        model_meta_data=None,
+    ):
+        self.manager = ScheduledModifierManager.from_yaml(SparseMLTrainer.recipe)
+
+        self.epochs = self.manager.max_epochs
+
+        if self.manager.learning_rate_modifiers:
+            lr_scheduler = None
+
+        self.manager.initialize(model)
+        optimizer = self.manager.modify(model, optimizer, num_batches_per_epoch)
+
+        super().__init__(
+            model,
+            loss,
+            optimizer,
+            out,
+            checkpoint_shell=checkpoint_shell,
+            lr_scheduler=lr_scheduler,
+            device=device,
+            model_meta_data=model_meta_data,
+        )
+
+    @classmethod
+    def cli(cls, parser: argparse.ArgumentParser):
+        openpifpaf.network.Trainer.cli(parser)
+        group = parser.add_argument_group("sparseml")
+        group.add_argument(
+            "--recipe", default=None, required=True, help="Path to sparseml recipe"
+        )
+
+    @classmethod
+    def configure(cls, args: argparse.Namespace):
+        openpifpaf.network.Trainer.configure(args)
+        cls.recipe = args.recipe
+
+    def loop(
+        self,
+        train_scenes: torch.utils.data.DataLoader,
+        val_scenes: torch.utils.data.DataLoader,
+        start_epoch=0,
+    ):
+        super().loop(train_scenes, val_scenes, start_epoch)
+        self.manager.finalize(self.model)
+
+    def train(self, scenes, epoch):
+        if self.manager.qat_active(epoch=epoch):
+            self.ema_restore_params = None
+        return super().train(scenes, epoch)
+
+    def write_model(self, epoch, final=True):
+        if torch.distributed.is_initialized() and torch.distributed.get_rank() != 0:
+            return
+
+        model_to_save = self.model
+        if self.checkpoint_shell is not None:
+            model = (
+                self.model if not hasattr(self.model, "module") else self.model.module
+            )
+            self.checkpoint_shell.load_state_dict(model.state_dict())
+            model_to_save = self.checkpoint_shell
+
+        filename = "{}.epoch{:03d}".format(self.out, epoch)
+        LOG.debug("about to write model")
+        torch.save(
+            {
+                "model": model_to_save,
+                "state_dict": model_to_save.state_dict(),
+                "checkpoint_recipe": str(self.manager),
+                "epoch": epoch,
+                "meta": self.model_meta_data,
+            },
+            filename,
+        )
+        LOG.info("model written: %s", filename)
+
+        if final:
+            sha256_hash = hashlib.sha256()
```

Comment:
Is this something we are doing now by default? Generating the model string from sha256?

Reply:
SHould probably make a not of this - I just copied this function from their trainer class and updated the stuff in the checkpoint dictionary. So this is their logic. We can probably change it to something else if we don't want to keep that